packages/flutter/lib/src/widgets/inherited_theme.dart

```diff
@@ -114,10 +114,18 @@ abstract class InheritedTheme extends InheritedWidget {
     assert(context != null);
 
     final List<InheritedTheme> themes = <InheritedTheme>[];
+    final Set<Type> themeTypes = <Type>{};
     context.visitAncestorElements((Element ancestor) {
       if (ancestor is InheritedElement && ancestor.widget is InheritedTheme) {
         final InheritedTheme theme = ancestor.widget;
-        themes.add(theme);
+        final Type themeType = theme.runtimeType;
```

Comment:
Rather than only recording one theme of each type, could you instead pass `themes.reversed` to `_CaptureAll(...)` below? That way it's a slightly more accurate recreation of the "`Theme`" tree.

I highly doubt this ever happens, but if you had `Theme(child: IconTheme(child: Theme(...)))`, and the `IconTheme` called `Theme.of(context)`, we'd still want it to return the outer theme.

Reply:
The theme wrapping does proceed inside-out (so to speak).  The source of the problem being fixed here was the InheritTheme wrap methods which look like:
```
  Widget wrap(BuildContext context, Widget child) {
    final IconTheme ancestorTheme = context.ancestorWidgetOfExactType(IconTheme);
    return identical(this, ancestorTheme) ? child : IconTheme(data: data, child: child);
  }
```
The identical test defends against the common case where the same inherited theme is visible in both the captureAll() and the wrap() contexts. When the same theme is visible in both contexts we don't bother wrapping a copy of the theme around the child. 

The source of the bug was the result of applying this test to an outer theme that was shadowed by an inner theme of the same type.  In that case the code was mistakenly wrapping the child with the outer theme because it wasn't identical to the inner one. Since the outer theme is shadowed by the inner one, we now just discard it in captureAll().

You're right that one could create an InheritedTheme whose build method created a dependency on the same type of theme. We definitely don't have"cascading" themes that work in this way although, currently we don't do anything to actively prevent it.